src/server/api/login.ts

```diff
@@ -1,13 +1,54 @@
 import { Router as createRouter } from 'express';
 import logger from '../logger';
+import { getUserByUsername } from '../database';
+import { isObjectRecord } from '../common/utilities/types';
+import { saltAndHash } from '../common/utilities/crypto';
 
 const router = createRouter();
 
 router.post('/', (req, res) => {
-  // TODO: implement this endpoint
-  logger.warning('Unimplemented endpoint: /api/login');
-  res.json({
-    success: false,
+  (async(): Promise<void> => {
+    if (!isObjectRecord(req.body)) {
+      throw new Error('api/login: req.body is not object');
+    }
+    const { username, password } = req.body;
+
+    if (typeof username !== 'string') {
+      throw new Error('api/login: user not type string');
+    }
+    if (typeof password !== 'string') {
+      throw new Error('api/login: password not type string');
+    }
+
+    const userObject = await getUserByUsername(username);
+
+    const { password: passwordDB, salt: saltDB }
+     = userObject;
+
+    const saltedAndHashedLoginPassword: Buffer = saltAndHash(
+      password,
+      Buffer.from(saltDB, 'hex')
+    );
+    const checkPassword = saltedAndHashedLoginPassword.toString('hex');
+
+    if (checkPassword === passwordDB) {
+      logger.info('Check: passwords are the same!');
+      // const tokenHashing = crypt.hmac('sha256', key);
+      // tokenValue = encodeToken();
+      // localStorage.setItem('token', JSON.stringify(token));
+    } else {
+      logger.info('FAILURE: Passwords not the same');
+    }
```

Comment:
```suggestion
    if (checkPassword !== passwordDB) {
      throw new Error('Unable to authenticate with the provided username and password');
    }
```

I generally recommend using a "Guard clause" for things like these: https://en.wikipedia.org/wiki/Guard_(computer_science)

The basic idea is that you want to exit early for the unexpected paths, either by returning early or throwing an error early. Then, we can continue the happy path.

For another simple example, rather than this:

```
if (validUsername) {
  if (validPassword) {
    res.json({ success: true });
  } else {
    throw new Error('Password is invalid.');
  }
} else {
  throw new Error('Username is invalid.');
}
```

We would add the guards at the top, and then proceed with the usual path:

```
if (!validUsername) {
  throw new Error('Username is invalid.');
}

if (!validPassword) {
  throw new Error('Password is invalid.');
}


res.json({ success: true });
```

Reply:
You already seem to be doing this well in other places 👍